R/method.R

```diff
@@ -1,145 +1,163 @@
+#' @param generic A generic function, either created by [new_generic()],
+#'   [new_external_generic()], or an existing S3 generic.
+#' @param signature A method signature. For R7 generics that use single
+#'   dispatch, this should be one of the following:
+#'   * An R7 class (created by [new_class()]).
+#'   * An R7 union (created by [new_union()]).
+#'   * An S3 class (created by [s3_class()]).
+#'   * An S4 class (created by [methods::getClass()] or [methods::new()]).
+#'   * A base type specified either with its constructor (`logical`, `integer`,
+#'     `double` etc) or its name (`"logical"`, `"integer"`, "`double`" etc).
+#'   * A base union type specified by its name: `"numeric"`, `"atomic"`, or
+#'     `"vector"`.
+#'
+#'   For R7 generics that use multiple dispatch, this can be a list of any of
+#'   the above types.
+#'
+#'   For S3 generics, this must be an R7 class.
 #' @param value A function that implements the generic specification for the
-#'   given `signature`. The arguments must be compatible with the generic.
-#' @importFrom utils getS3method
+#'   given `signature`.
 #' @export
 #' @examples
 #' # Create a generic
 #' bizarro <- new_generic("bizarro", dispatch_args = "x")
 #' # Register some methods
-#' method(bizarro, "numeric") <- function(x, ...) rev(x)
-#' method(bizarro, s3_class("factor")) <- function(x, ...) {
-#'   levels(x) <- rev(levels(x))
-#'   x
-#' }
-#' method(bizarro, s3_class("data.frame")) <- function(x, ...) {
+#' method(bizarro, "numeric") <- function(x) rev(x)
+#' method(bizarro, s3_class("data.frame")) <- function(x) {
 #'   x[] <- lapply(x, bizarro)
 #'   rev(x)
 #' }
 #'
 #' # Using a generic calls the methods automatically
-#' bizarro(1)
-#'
-#' # But it can be useful to explicitly retrieve a method in order to
-#' # inspect its implementation
-#' method(bizarro, "double")
-#' method(bizarro, s3_class("factor"))
-method <- function(generic, signature) {
-  # TODO: check that signature doesn't contain any unions
-
-  signature <- as_signature(signature)
-  method_impl(generic, signature, ignore = NULL)
+#' bizarro(head(mtcars))
+`method<-` <- function(generic, signature, value) {
+  register_method(generic, signature, value)
+  invisible(generic)
 }
 
-methods <- function(generic) {
-  get_all_methods(generic@methods, character())
-}
+register_method <- function(generic, signature, method, package = packageName(parent.frame())) {
+  signature <- as_signature(signature)
+  generic <- as_generic(generic)
 
-get_all_methods <- function(x, signature) {
-  if(!is.environment(x)) {
-    return(x)
+  if (is_external_generic(generic)) {
+    if (!is.null(package)) {
+      # method registration within package, so add to lazy registry
+      external_methods_add(package, generic, signature, method)
+    } else {
+      # otherwise find the generic and register
+      generic <- getFromNamespace(generic$name, asNamespace(generic$package))
+      register_method(generic, signature, method)
```

Comment:
I think the fact that `package` was always coming out as `"R7"` (mentioned over Slack) implies that we don't have a test for this branch?